File: 프로그래머스/lv2_프린터.cpp
#include <string>
#include <vector>

using namespace std;

int solution(vector<int> priorities, int location) {
    int answer = 0;
    bool re_push = false;

    vector<pair<int, int>> v;
    pair<int, int> cur;
    
    // location이 포함된 새로운 벡터 생성
    for( int i=0; i<priorities.size(); i++){
        v.push_back(make_pair(priorities[i],i));
    }

    while(1){
        cur = v.front();
        v.erase(v.begin());
        re_push = false;

        for( int j=0; j<v.size(); j++){
            // 원소중 현재 선택된 원소의 우선순위보다 큰게 있다면 마지막으로 푸쉬
            if(cur.first < v[j].first ){
                v.push_back(cur);
                re_push = true;
                break;
            }
        }
        // 현재 선택된 원소보다 우선순위가 큰 원소가 없다면
        if( re_push == false ){
            answer++;
            if( re_push == false && cur.second == location){
                return answer;
            }
        }
    }


    return answer;
}

int main(void){

    vector<int> priorities = {1,1,9,1,1,1};
    int location = 0;
    int ret;

    ret = solution(priorities, location);

    printf("%d\n", ret);

    return 0;
}  

File: 프로그래머스/lv2_프린터.py
from collections import deque

'''
    location을 어떻게 관리할까 고민하다가, 계속 변경해주는 방식으로 해결했는데,
    enumerate를 사용해서 인덱스를 미리 저장해놓고 사용하면, 그럴 필요가 없어짐.
'''

def solution( prioirities, location):
    answer = 0
    cnt = 0
    deq = deque(prioirities)

    while( location != -1 ):
        if( deq[0] != max(deq)):
            deq.append(deq.popleft())
            if location == 0:
                location = len(deq)-1
            else:
                location -= 1
        else:
            deq.popleft()
            cnt += 1
            location -= 1

        # print(deq, location)

    answer = cnt

    return answer

# def solution(priorities, location):
#     queue =  [(i,p) for i,p in enumerate(priorities)]
#     answer = 0
#     while True:
#         cur = queue.pop(0)
#         if any(cur[1] < q[1] for q in queue):
#             queue.append(cur)
#         else:
#             answer += 1
#             if cur[0] == location:
#                 return answer


if __name__ == "__main__":
    ret = solution([2, 1, 3, 2], 2)
    print(ret)

    ret = solution([1, 1, 9, 1, 1, 1], 0)
    print(ret)